go/arrow/schema.go

```diff
@@ -112,9 +112,10 @@ func NewSchema(fields []Field, metadata *Metadata) *Schema {
 	return sc
 }
 
-func (sc *Schema) Metadata() Metadata { return sc.meta }
-func (sc *Schema) Fields() []Field    { return sc.fields }
-func (sc *Schema) Field(i int) Field  { return sc.fields[i] }
+func (sc *Schema) Metadata() Metadata           { return sc.meta }
+func (sc *Schema) Fields() []Field              { return sc.fields }
+func (sc *Schema) Field(i int) Field            { return sc.fields[i] }
+func (sc *Schema) SetFieldName(i int, n string) { sc.fields[i].Name = n }
```

Comment:
done like this, this breaks the `FieldByName` method as we haven't updated the `index` field of `Schema`.

perhaps the easiest (and preserve the "functional, immutable" aspect of `Schema`) way is to get rid of this `SetFieldName` method and just recreate a new Schema (with the new Field name) in the `Reader.readHeader` method.